Quarkus Maven plugin is unable to resolve local deps when generating code tests

### Describe the bug

https://github.com/quarkusio/quarkus/pull/34286 broke build of our tests with code test generation.

### Expected behavior

Quarkus Maven plugin should resolve artifacts from local repo. It worked with 3.1.3.Final.

### Actual behavior

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:generate-code-tests (default) on project examples-quarkus-helm: Quarkus code generation phase has failed: Failed to bootstrap application in TEST mode: Failed to resolve dependencies for io.quarkus.qe:examples-quarkus-helm:jar:1.3.1.Beta21-SNAPSHOT: Failed to collect dependencies at io.quarkus.qe:quarkus-test-core:jar:1.3.1.Beta21-SNAPSHOT: Failed to read artifact descriptor for io.quarkus.qe:quarkus-test-core:jar:1.3.1.Beta21-SNAPSHOT: Could not transfer artifact io.quarkus.qe:quarkus-test-core:pom:1.3.1.Beta21-SNAPSHOT from/to gradle-official-repository (https://repo.gradle.org/gradle/libs-releases-local/): transfer failed for https://repo.gradle.org/gradle/libs-releases-local/io/quarkus/qe/quarkus-test-core/1.3.1.Beta21-SNAPSHOT/quarkus-test-core-1.3.1.Beta21-SNAPSHOT.pom, status: 409 Conflict -> [Help 1]
```

### How to Reproduce?

Reproducer:

```bash
git clone git@github.com:quarkus-qe/quarkus-test-framework.git
cd quarkus-test-framework
# only for repeated steps, it is important to have empty .m2 QE artifacts
sudo rm ~/.m2/repository/io/quarkus/qe -R
mvn -V -B -s .github/mvn-settings.xml verify -Pframework,examples -Dvalidate-format -DskipTests -DskipITs
```

### Output of `uname -a` or `ver`

Fedora

### Output of `java -version`

OpenJDK Runtime Environment Temurin-17.0.7+7

### GraalVM version (if different from Java)

22.3.2

### Quarkus version or git rev

999-SNAPSHOT, 3.2.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.8 and 3.9.3

### Additional information

_No response_

--- a/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/DependencyInProfileActiveByDefaultRawModelBuilderTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/DependencyInProfileActiveByDefaultRawModelBuilderTest.java
@@ -96,8 +96,8 @@ public class DependencyInProfileActiveByDefaultRawModelBuilderTest extends Boots
         assertThat(d).isNotNull();
         assertThat(d.isRuntimeCp()).isTrue();
         assertThat(d.isDeploymentCp()).isTrue();
-        assertThat(d.isWorkspaceModule()).isFalse(); // limitation of the raw pom-based workspace discovery
-        assertThat(d.isReloadable()).isFalse();
+        assertThat(d.isWorkspaceModule()).isTrue();
+        assertThat(d.isReloadable()).isTrue();
         assertThat(d.isRuntimeExtensionArtifact()).isFalse();
 
         d = deps.get("common-library");
@@ -105,7 +105,7 @@ public class DependencyInProfileActiveByDefaultRawModelBuilderTest extends Boots
         assertThat(d.isRuntimeCp()).isTrue();
         assertThat(d.isDeploymentCp()).isTrue();
         assertThat(d.isWorkspaceModule()).isTrue();
-        assertThat(d.isReloadable()).isFalse(); // since it's a dependency of a non-reloadable module
+        assertThat(d.isReloadable()).isTrue();
         assertThat(d.isRuntimeExtensionArtifact()).isFalse();
 
         d = deps.get("ext-a");

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/WorkspaceLoader.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/WorkspaceLoader.java
@@ -219,8 +219,22 @@ public class WorkspaceLoader implements WorkspaceModelResolver, WorkspaceReader 
     }
 
     private LocalProject loadProjectModules(LocalProject project, String skipModule) throws BootstrapMavenException {
-        final List<String> modules = project.getModelBuildingResult() == null ? project.getRawModel().getModules()
-                : project.getModelBuildingResult().getEffectiveModel().getModules();
+        final List<String> modules;
+        if (project.getModelBuildingResult() == null) {
+            var projectModel = project.getRawModel();
+            List<String> combinedList = null;
+            for (var profile : projectModel.getProfiles()) {
+                if (!profile.getModules().isEmpty()) {
+                    if (combinedList == null) {
+                        combinedList = new ArrayList<>(projectModel.getModules());
+                    }
+                    combinedList.addAll(profile.getModules());
+                }
+            }
+            modules = combinedList == null ? projectModel.getModules() : combinedList;
+        } else {
+            modules = project.getModelBuildingResult().getEffectiveModel().getModules();
+        }
         if (!modules.isEmpty()) {
             for (String module : modules) {
                 if (module.equals(skipModule)) {

--- a/independent-projects/bootstrap/maven-resolver/src/test/java/io/quarkus/bootstrap/workspace/test/LocalWorkspaceDiscoveryTest.java
+++ b/independent-projects/bootstrap/maven-resolver/src/test/java/io/quarkus/bootstrap/workspace/test/LocalWorkspaceDiscoveryTest.java
@@ -204,11 +204,11 @@ public class LocalWorkspaceDiscoveryTest {
     }
 
     @Test
-    public void loadModulesInProfiles() throws Exception {
+    public void loadEffectiveModelBuilderModulesInProfiles() throws Exception {
         final URL moduleUrl = Thread.currentThread().getContextClassLoader()
                 .getResource("modules-in-profiles/integration-tests/rest-tests");
         assertNotNull(moduleUrl);
-        final Path moduleDir = Paths.get(moduleUrl.toURI());
+        final Path moduleDir = Path.of(moduleUrl.toURI());
 
         final LocalProject module1 = new BootstrapMavenContext(BootstrapMavenContext.config()
                 .setEffectiveModelBuilder(true)
@@ -223,6 +223,28 @@ public class LocalWorkspaceDiscoveryTest {
         assertNotNull(ws.getProject("org.acme", "acme-integration-tests"));
         assertNotNull(ws.getProject("org.acme", "acme-rest-tests"));
         assertEquals(6, ws.getProjects().size());
+    }
+
+    @Test
+    public void loadModulesInProfiles() throws Exception {
+        final URL moduleUrl = Thread.currentThread().getContextClassLoader()
+                .getResource("modules-in-profiles/integration-tests/rest-tests");
+        assertNotNull(moduleUrl);
+        final Path moduleDir = Path.of(moduleUrl.toURI());
+
+        final LocalProject module1 = new BootstrapMavenContext(BootstrapMavenContext.config()
+                .setCurrentProject(moduleDir.toString()))
+                .getCurrentProject();
+        final LocalWorkspace ws = module1.getWorkspace();
+
+        assertNotNull(ws.getProject("org.acme", "quarkus-quickstart-multimodule-parent"));
+        assertNotNull(ws.getProject("org.acme", "quarkus-quickstart-multimodule-html"));
+        assertNotNull(ws.getProject("org.acme", "quarkus-quickstart-multimodule-main"));
+        assertNotNull(ws.getProject("org.acme", "quarkus-quickstart-multimodule-rest"));
+        assertNotNull(ws.getProject("org.acme", "acme-integration-tests"));
+        assertNotNull(ws.getProject("org.acme", "acme-rest-tests"));
+        assertNotNull(ws.getProject("org.acme", "other"));
+        assertEquals(7, ws.getProjects().size());
     }
 
     @Test

--- a/independent-projects/bootstrap/maven-resolver/src/test/resources/modules-in-profiles/other/pom.xml
+++ b/independent-projects/bootstrap/maven-resolver/src/test/resources/modules-in-profiles/other/pom.xml
@@ -1,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project>
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.acme</groupId>
+    <artifactId>other</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+</project>

--- a/independent-projects/bootstrap/maven-resolver/src/test/resources/modules-in-profiles/pom.xml
+++ b/independent-projects/bootstrap/maven-resolver/src/test/resources/modules-in-profiles/pom.xml
@@ -43,5 +43,11 @@
                 <module>rest</module>
             </modules>
         </profile>
+        <profile>
+            <id>other</id>
+            <modules>
+                <module>other</module>
+            </modules>
+        </profile>
     </profiles>
 </project>

